File: atcoder/abc213_e_01bfs.py
# https://atcoder.jp/contests/ABC213/tasks/abc213_e

# import heapq
from collections import deque
dy = [-1, 1, 0, 0]
dx = [ 0, 0, 1,-1]
py = [-2,-2,-2,-1,-1,-1,-1,-1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2, 2]
px = [-1, 0, 1,-2,-1, 0, 1, 2,-2,-1, 1, 2,-2,-1, 0, 1, 2,-1, 0, 1]
INF = 1001001001

H, W = map(int, input().split())
map = []
dist = [[INF]*W for _ in range(H)]
for i in range(H):
    S = input().rstrip()
    map.append(S)
# h = []
# heapq.heappush(h, (0, 0, 0))
q = deque([(0,0,0)])
ans = 0
while len(q):
    # d, y, x = heapq.heappop(h)
    d, y, x = q.pop()
    if y==H-1 and x==W-1:
        ans = d
        break
    for i in range(4):
        ny = y + dy[i]
        if ny<0 or ny>=H: continue
        nx = x + dx[i]
        if nx<0 or nx>=W: continue
        if map[ny][nx]=='#': continue
        if d < dist[ny][nx]:
            # heapq.heappush(h, (d, ny, nx))
            q.append((d, ny, nx))
            dist[ny][nx] = d
    for i in range(20):
        ny = y + py[i]
        if ny<0 or ny>=H: continue
        nx = x + px[i]
        if nx<0 or nx>=W: continue
        nd = d + 1
        if nd < dist[ny][nx]:
            # heapq.heappush(h, (d+1, ny, nx))
            q.appendleft((nd, ny, nx))
            dist[ny][nx] = nd
print(ans)

File: atcoder/abc213_e_01bfs.cpp
// https://atcoder.jp/contests/abc213/tasks/abc213_e
// https://atcoder.jp/contests/abc213/submissions/24849890
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
// #include <atcoder/all>
// using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

const int di[] = {-1,0,1,0};
const int dj[] = {0,-1,0,1};

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
	int h, w;
	cin >> h >> w;
    vector<string> s(h);
    rep(i,h) cin >> s[i];
    deque<P> q;
    vector dist(h, vector<int>(w, INF));
    vector visited(h, vector<bool>(w));

    dist[0][0] = 0;
    q.emplace_back(0,0);
    while (q.size()) {
        auto [i,j] = q.front(); q.pop_front();
        if (visited[i][j]) continue;
        visited[i][j] = true;
        int d = dist[i][j];
        rep(v,4) {
            int ni = i+di[v], nj = j+dj[v];
            if (ni<0 || nj<0 || ni>=h || nj>=w) continue;
            if (s[ni][nj]=='#') continue;
            if (dist[ni][nj]<=d) continue;
            dist[ni][nj] = d;
            q.emplace_front(ni,nj);
        }
        for (int ei = -2; ei <= 2; ++ei) {
            for (int ej = -2; ej <= 2; ++ej) {
                if (abs(ei)+abs(ej)>3) continue;
                int ni = i+ei, nj = j+ej;
                if (ni<0 || nj<0 || ni>=h || nj>=w) continue;
                if (dist[ni][nj]<=d+1) continue;
                dist[ni][nj] = d+1;
                q.emplace_back(ni,nj);
            }
        }
    }
    cout << dist[h-1][w-1] << endl;
	return 0;
}
